R/getReportMAgPIE2REMIND.R

```diff
@@ -0,0 +1,133 @@
+#'   Toplevel|+|Item 1
+#'   Toplevel|+|Item 2
+#'   Toplevel|Item 2|+|Subitem 1
+#'   Toplevel|Item 2|+|Subitem 1
+#'   Toplevel|++|Item A
+#'   Toplevel|++|Item B
+#'   Toplevel|Item ?}
+#'
+#' @author Florian Humpenoeder, David Klein
+#' @importFrom magclass write.report2 getSets add_dimension
+#' @importFrom methods is
+#' @examples
+#' \dontrun{
+#' x <- getReportMAgPIE2REMIND(gdx)
+#' }
+#'
+getReportMAgPIE2REMIND <- function(gdx, file = NULL, scenario = NULL, filter = c(1, 2, 7),
+                      detail = TRUE, dir = ".", ...) {
+  tryReport <- function(report, width, gdx) {
+    regs <- c(readGDX(gdx, "i"), "GLO")
+    years <- readGDX(gdx, "t")
+    message("   ", format(report, width = width), appendLF = FALSE)
+    t <- system.time(x <- try(eval(parse(text = paste0("suppressMessages(", report, ")"))), silent = TRUE))
+    t <- paste0(" ", format(t["elapsed"], nsmall = 2, digits = 2), "s")
+    if (is(x, "try-error")) {
+      message("ERROR", t)
+      x <- NULL
+    } else if (is.null(x)) {
+      message("no return value", t)
+      x <- NULL
+    } else if (is.character(x)) {
+      message(x, t)
+      x <- NULL
+    } else if (!is.magpie(x)) {
+      message("ERROR - no magpie object", t)
+      x <- NULL
+    } else if (!setequal(getYears(x), years)) {
+      message("ERROR - wrong years", t)
+      x <- NULL
+    } else if (!setequal(getItems(x, dim = 1), regs)) {
+      message("ERROR - wrong regions", t)
+      x <- NULL
+    } else if (any(grepl(".", getNames(x), fixed = TRUE))) {
+      message("ERROR - data names contain dots (.)", t)
+      x <- NULL
+    } else {
+      message("success", t)
+    }
+    return(x)
+  }
+
+  tryList <- function(..., gdx) {
+    width <- max(nchar(c(...))) + 1
+    return(lapply(unique(list(...)), tryReport, width, gdx))
+  }
+
+  message("Start getReportMAgPIE2REMIND(gdx)...")
+
+  t <- system.time(
+    output <- tryList("reportDemandBioenergy(gdx,detail=detail)",
+      "reportEmissions(gdx)",
+      "reportCosts(gdx)",
+      "reportPriceBioenergy(gdx)",
+      gdx = gdx
+    )
+  )
+
+  message(paste0("Total runtime:  ", format(t["elapsed"], nsmall = 2, digits = 2), "s"))
+
+  output <- .filtermagpie(mbind(output), gdx, filter = filter)
+
+  getSets(output, fulldim = FALSE)[3] <- "variable"
+
+  if (!is.null(scenario)) {
+    output <- add_dimension(output,
+      dim = 3.1,
+      add = "scenario",
+      nm = gsub(".", "_", scenario, fixed = TRUE)
+    )
+  }
+  output <- add_dimension(output, dim = 3.1, add = "model", nm = "MAgPIE")
+
+  missingUnit <- !grepl("\\(.*\\)", getNames(output))
+  if (any(missingUnit)) {
+    warning("Some units are missing in getReportMAgPIE2REMIND!")
+    warning("Missing units in:", getNames(output)[which(!grepl("\\(.*\\)", getNames(output)) == TRUE)])
+    getNames(output)[missingUnit] <- paste(getNames(output)[missingUnit], "( )")
+  }
+  if (!is.null(file)) {
+    write.report2(output, file = file, ...)
+  } else {
+    return(output)
+  }
```

Comment:
Regarding your proposal using `invisible` I only see this these lines to use it for:
```suggestion
  if (!is.null(file)) write.report(output, file = file, ...)
  return(invisible(output))
```

Reply:
True, and most of the savings just come from cramming the if statement on one line which I really don't like. I guess I just don't like conditional returns, because R will return something implicitly anyway (in this case it would be the result of write.report2). Just returning invisibly would make the documentation slightly easier. In a hypothetical case where someone wants to write the report to a file but also wants to work further with the report in R it might make sense to return anyway. So all in all I let you decide of course, my opinion on this is not as strong as it might seem. Returning invisibly is often just a no-downside way of doing it.